src/linear_algebra/lagrange.lean

```diff
@@ -0,0 +1,106 @@
+  that evaluates to `f x` at `x` for `x ∈ s`.
+
+-/
+
+noncomputable theory
+open_locale classical
+
+universe u
+
+namespace lagrange
+
+variables {F : Type u} [decidable_eq F] [field F] (s : finset F)
+
+open polynomial
+
+/-- Lagrange basis polynomials that evaluate to 1 at `x` and 0 at other elements of `s`. -/
+def basis (x : F) : polynomial F :=
+(s.erase x).prod $ λ y, C (x - y)⁻¹ * (X - C y)
+
+@[simp] theorem basis_empty (x : F) : basis ∅ x = 1 :=
+rfl
+
+@[simp] theorem eval_basis_self (x : F) : (basis s x).eval x = 1 :=
+begin
+  rw [basis, ← (s.erase x).prod_hom (eval x), finset.prod_eq_one],
+  intros y hy, simp_rw [eval_mul, eval_sub, eval_C, eval_X],
+  exact inv_mul_cancel (sub_ne_zero_of_ne (finset.ne_of_mem_erase hy).symm)
+end
+
+@[simp] theorem eval_basis_ne (x y : F) (h1 : y ∈ s) (h2 : y ≠ x) : (basis s x).eval y = 0 :=
+begin
+  rw [basis, ← (s.erase x).prod_hom (eval y), finset.prod_eq_zero (finset.mem_erase.2 ⟨h2, h1⟩)],
+  simp_rw [eval_mul, eval_sub, eval_C, eval_X, sub_self, mul_zero]
+end
+
+theorem eval_basis (x y : F) (h : y ∈ s) : (basis s x).eval y = if y = x then 1 else 0 :=
+by { split_ifs with H, { subst H, apply eval_basis_self }, { exact eval_basis_ne s x y h H } }
+
+@[simp] theorem nat_degree_basis (x : F) (hx : x ∈ s) : (basis s x).nat_degree = s.card - 1 :=
+begin
+  unfold basis, generalize hsx : s.erase x = sx,
+  have : x ∉ sx := hsx ▸ finset.not_mem_erase x s,
+  rw [← finset.insert_erase hx, hsx, finset.card_insert_of_not_mem this, nat.add_sub_cancel],
+  clear hx hsx s, revert this, apply sx.induction_on,
+  { intros hx, rw [finset.prod_empty, nat_degree_one], refl },
+  { intros y s hys ih hx, rw [finset.mem_insert, not_or_distrib] at hx,
+    have h1 : C (x - y)⁻¹ ≠ C 0 := λ h, hx.1 (eq_of_sub_eq_zero $ inv_eq_zero.1 $ C_inj.1 h),
+    have h2 : X ^ 1 - C y ≠ 0 := by convert X_pow_sub_C_ne_zero zero_lt_one _,
+    rw C_0 at h1, rw pow_one at h2,
+    rw [finset.prod_insert hys, nat_degree_mul_eq (mul_ne_zero h1 h2), ih hx.2,
+        finset.card_insert_of_not_mem hys, nat_degree_mul_eq h1 h2,
+        nat_degree_C, zero_add, nat_degree, degree_X_sub_C, add_comm], refl,
+    rw [ne, finset.prod_eq_zero_iff], rintro ⟨z, hzs, hz⟩,
+    rw mul_eq_zero at hz, cases hz with hz hz,
+    { rw [← C_0, C_inj, inv_eq_zero, sub_eq_zero] at hz, exact hx.2 (hz.symm ▸ hzs) },
+    { rw ← pow_one (X : polynomial F) at hz, exact X_pow_sub_C_ne_zero zero_lt_one _ hz } }
+end
+
+variables (f : (↑s : set F) → F)
+
+/-- Lagrange interpolation: given a finset `s` and a function `f : s → F`,
+`interpolate s f` is the unique polynomial of degree `< s.card`
+that takes value `f x` on all `x` in `s`. -/
+def interpolate : polynomial F :=
+s.attach.sum $ λ x, C (f x) * basis s x
+
+@[simp] theorem interpolate_empty (f) : interpolate (∅ : finset F) f = 0 :=
+rfl
+
+@[simp] theorem eval_interpolate (x ∈ s) : eval x (interpolate s f) = f ⟨x, H⟩ :=
```

Comment:
```suggestion
@[simp] theorem eval_interpolate (H : x ∈ s) : eval x (interpolate s f) = f ⟨x, H⟩ :=
```

Reply:
`(x ∈ s)` is short hand for `(x) (H : x ∈ s)`

Reply:
Yes, but I don't think we should rely on auto-generated names if we can avoid it.

Reply:
I'm just pointing out that your suggestion wouldn't compile.

Reply:
Oh, silly me.
```suggestion
@[simp] theorem eval_interpolate (x) (H : x ∈ s) : eval x (interpolate s f) = f ⟨x, H⟩ :=
```